Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -38,59 +43,220 @@ open scoped Matrix
+    simp [← pow_two, ← IsROrC.ofReal_pow, Real.sq_sqrt (hA.eigenvalues_nonneg v)]
+  rw [this]
+  convert hA.1.spectral_theorem'.symm
+  apply Matrix.IsHermitian.conjTranspose_eigenvectorMatrix
+
+/-- If `A, B` are positive semidefinite and `A ^ 2 = B ^ 2`, then `A = B`. -/
+lemma eq_of_sq_eq_sq {B : Matrix n n 𝕜} (hB : PosSemidef B) (hAB : A ^ 2 = B ^ 2) : A = B := by
+  /- This is deceptively hard, much more difficult than the positive *definite* case. We follow a
+  clever proof due to Koeber and Schäfer. The idea is that if `A ≠ B`, then `A - B` has a nonzero
+  real eigenvalue, with eigenvector `v`. Then a manipulation using the identity
+  `A ^ 2 - B ^ 2 = A * (A - B) + (A - B) * B` leads to the conclusion that
+  `⟨v, A v⟩ + ⟨v, B v⟩ = 0`. Since `A, B` are positive semidefinite, both terms must be zero. Thus
+  `⟨v, (A - B) v⟩ = 0`, but this is a nonzero scalar multiple of `⟨v, v⟩`, contradiction. -/
+  by_contra h_ne
+  let ⟨v, t, ht, hv, hv'⟩ := (hA.1.sub hB.1).exists_eigenvector_of_ne_zero (sub_ne_zero.mpr h_ne)
+  have h_sum : 0 = t * (star v ⬝ᵥ mulVec A v + star v ⬝ᵥ mulVec B v)
+  · calc
+      0 = star v ⬝ᵥ mulVec (A ^ 2 - B ^ 2) v := by rw [hAB, sub_self, zero_mulVec, dotProduct_zero]
+      _ = star v ⬝ᵥ mulVec A (mulVec (A - B) v)  + star v ⬝ᵥ mulVec (A - B) (mulVec B v) := by
+        rw [mulVec_mulVec, mulVec_mulVec, ← dotProduct_add, ← add_mulVec, mul_sub, sub_mul,
+          add_sub, sub_add_cancel, pow_two, pow_two]
+      _ = t * (star v ⬝ᵥ mulVec A v) + vecMul (star v) (A - B)ᴴ ⬝ᵥ mulVec B v := by
+        rw [hv', mulVec_smul, dotProduct_smul, IsROrC.real_smul_eq_coe_mul,
+          dotProduct_mulVec _ (A - B), hA.1.sub hB.1]
+      _ = t * (star v ⬝ᵥ mulVec A v + star v ⬝ᵥ mulVec B v) := by
+        simp_rw [← star_mulVec, hv', mul_add, ← IsROrC.real_smul_eq_coe_mul, ← smul_dotProduct]
+        congr 2 with i
+        simp only [Pi.star_apply, Pi.smul_apply, IsROrC.real_smul_eq_coe_mul, star_mul',
+          IsROrC.star_def, IsROrC.conj_ofReal]
+  replace h_sum : star v ⬝ᵥ mulVec A v + star v ⬝ᵥ mulVec B v = 0
+  · rw [eq_comm, ← mul_zero (t : 𝕜)] at h_sum
+    exact mul_left_cancel₀ (IsROrC.ofReal_ne_zero.mpr ht) h_sum
+  have h_van : star v ⬝ᵥ mulVec A v = 0 ∧ star v ⬝ᵥ mulVec B v = 0
+  · refine ⟨le_antisymm ?_ (hA.2 v), le_antisymm ?_ (hB.2 v)⟩
+    · rw [add_comm, add_eq_zero_iff_eq_neg] at h_sum
+      simpa only [h_sum, neg_nonneg] using hB.2 v
+    · simpa only [add_eq_zero_iff_eq_neg.mp h_sum, neg_nonneg] using hA.2 v
+  have aux : star v ⬝ᵥ mulVec (A - B) v = 0
+  · rw [sub_mulVec, dotProduct_sub, h_van.1, h_van.2, sub_zero]
+  rw [hv', dotProduct_smul, IsROrC.real_smul_eq_coe_mul, ← mul_zero ↑t] at aux
+  exact hv <| Matrix.dotProduct_star_self_eq_zero.mp <| mul_left_cancel₀
+    (IsROrC.ofReal_ne_zero.mpr ht) aux
+
+lemma eq_sqrt_of_sq_eq {B : Matrix n n 𝕜} (hB : PosSemidef B) (hAB : A ^ 2 = B) : A = hB.sqrt :=
+  hA.eq_of_sq_eq_sq hB.posSemidef_sqrt (hB.sq_sqrt.symm ▸ hAB)
```

Comment:
I think it would be good to also show:
```
+lemma sqrt_sq (hA2 : PosSemidef (A ^ 2)) : (hA.).sqrt = A :=
+  hA2.posSemidef_sqrt.eq_of_sq_eq_sq hA hA2.sq_sqrt
```
```suggestion
lemma sqrt_sq (hA2 : PosSemidef (A ^ 2)) : hA2.sqrt = A :=
  hA2.posSemidef_sqrt.eq_of_sq_eq_sq hA hA2.sq_sqrt
```

Reply:
@Eric, I appreciate your feedback, but this constant ping-pong of tiny addition after tiny addition is getting to be a little frustrating. 

Could I ask that you perhaps make a finite list of the additional functionality you'd consider "sufficient" for getting this PR merged, and that once this list is made, you endeavour to resist the temptation to add to it?

Reply:
This PR is, after all, an outgrowth of#8594, based on code by my student who encountered Lean for the first time this semester. I encouraged him to make a PR because I wanted to demonstrate to him, and the rest of my class at ETH Zurich, that mathlib was a project that welcomed new contributors. We seem to be in some danger of demonstrating the opposite of that.